server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -770,7 +770,23 @@ public static Engine.Index prepareIndex(DocumentMapper docMapper,
                                 ifSeqNo, ifPrimaryTerm);
     }
 
-    private Engine.IndexResult index(Engine engine, Engine.Index index) throws IOException {
+    public Engine.IndexResult index(Engine.Index index) throws IOException {
+        assert index.primaryTerm() <= getOperationPrimaryTerm()
+                : "op term [ " + index.primaryTerm() + " ] > shard term [" + getOperationPrimaryTerm() + "]";
+        ensureWriteAllowed(index.origin());
+        try {
+            return index(getEngine(), index);
```

Comment:
very minor: can't we skip the `Engine` arg and call `getEngine` in the `index()` method?

Reply:
The method this is calling into is currently also called from other places where the engine is retrieved earlier.